File: baml_language/sdk_tests/crates/typescript/host_reflect/customizable/host_reflect.test.ts
import { BamlError, BamlType } from "@boundaryml/baml-bridge";
import { describe, expect, it } from "vitest";

import {
  HostParse,
  HostTypeEqual,
  HostTypeName,
  StaticChoice,
  StaticNamed,
  StaticRecord,
  reflect,
} from "./baml_sdk/index.js";

describe("BEP-066 TypeScript host boundary", () => {
  it("runtime_enum_definition_decodes_alias", () => {
    const red = reflect.enum.value("RED", { alias: "k7", description: "warm" });
    const category = reflect.enum.new("Category", [red, "BLUE"]);

    expect(category).toBeInstanceOf(BamlType);
    expect(HostParse<string>('"k7"', { $types: { T: category } })).toBe("RED");
  });

  it("runtime_class_definition_preserves_nested_metadata", () => {
    const category = reflect.enum.new("Category", [
      reflect.enum.value("RED", { alias: "k7" }),
    ]);
    const record = reflect.class.new("RuntimeRecord", {
      label: reflect.type.of(String).meta({ alias: "display_label" }),
      category,
      scores: reflect.type.of("int").array(),
    });

    const parsed = HostParse<Record<string, unknown>>(
      '{"display_label":"ok","category":"k7","scores":[1,2]}',
      { $types: { T: record } },
    );
    expect(parsed).toEqual({ label: "ok", category: "RED", scores: [1, 2] });
    expect(Object.getPrototypeOf(parsed)).toBeNull();
    expect(Object.keys(parsed)).toEqual(["label", "category", "scores"]);
  });

  it("compiled_package_returns_class_graph", () => {
    const pkg = reflect.Package.compile({
      "runtime.baml": "class CompiledRow { amount int note string? }",
    });
    const compiled = pkg.get_class("CompiledRow");

    expect(compiled).toBeInstanceOf(BamlType);
    expect(
      HostParse<Record<string, unknown>>('{"amount":7}', {
        $types: { T: compiled! },
      }),
    ).toEqual({ amount: 7, note: null });
  }, 30_000);

  it("wire_occurrences_are_fresh_and_handles_reject_serialization", () => {
    const runtimeType = reflect.class.new("Fresh", {
      value: reflect.type.of("int"),
    });
    expect(HostTypeEqual({ $types: { A: runtimeType, B: runtimeType } })).toBe(false);
    expect(() => JSON.stringify(runtimeType)).toThrow(/cannot be serialized/);
  });

  it("known_type_tokens_compose_and_reject_unknowns", () => {
    expect(HostTypeName({ $types: { T: String } })).toBe("string");
    expect(HostTypeName({ $types: { T: { list: { optional: String } } } })).toBe(
      "(string | null)[]",
    );
    expect(HostTypeName({ $types: { T: StaticChoice } })).toBe("StaticChoice");
    expect(HostTypeName({ $types: { T: StaticNamed } })).toBe("StaticNamed");
    expect(reflect.type.of("int").optional().array()).toBeInstanceOf(BamlType);
    expect(reflect).toHaveProperty("class");
    expect(reflect).toHaveProperty("type");

    class NotGenerated {}
    expect(() => HostTypeName({ $types: { T: NotGenerated } })).toThrow(
      /unsupported TypeScript type token/,
    );
  });

  it("generated_class_subclasses_resolve_to_declared_type", () => {
    class ChildRecord extends StaticRecord {}

    expect(HostTypeName({ $types: { T: ChildRecord } })).toBe("StaticRecord");
  });

  it("host_handles_expose_composition_only", () => {
    const runtimeType = reflect.type.of("int");
    expect("kind" in runtimeType).toBe(false);
    expect("fields" in runtimeType).toBe(false);
    expect("as_type" in runtimeType).toBe(false);
  });

  it("reflection_compile_errors_are_typed", () => {
    let thrown: unknown;
    try {
      reflect.Package.compile({ "broken.baml": "class {" });
    } catch (error) {
      thrown = error;
    }
    expect(thrown).toBeInstanceOf(BamlError);
    expect((thrown as BamlError).className).toBe("baml.reflect.errors.CompilationError");
  });
});

File: baml_language/sdk_tests/crates/go/host_reflect/customizable/host_reflect_test.go
package sdk_test

import (
	"bytes"
	"context"
	"encoding/gob"
	"encoding/json"
	"errors"
	stdreflect "reflect"
	"strings"
	"testing"

	"baml.local/sdk/baml_sdk"
	bamlreflect "baml.local/sdk/baml_sdk/reflect"
	baml_go "github.com/boundaryml/baml-go"
)

func Test_runtime_enum_definition_decodes_alias(t *testing.T) {
	ctx := context.Background()
	category, err := bamlreflect.Enum(ctx, "Category", []bamlreflect.EnumValue{
		bamlreflect.NewEnumValue("RED", bamlreflect.WithAlias("k7"), bamlreflect.WithDescription("warm")),
		bamlreflect.NewEnumValue("BLUE"),
	})
	if err != nil {
		t.Fatal(err)
	}
	got, err := baml_sdk.HostParse[any](ctx, `"k7"`, baml_go.WithTypeArg("T", category))
	if err != nil || got != "RED" {
		t.Fatalf("runtime enum parse = %#v, %v", got, err)
	}
}

func Test_runtime_class_definition_preserves_nested_metadata(t *testing.T) {
	ctx := context.Background()
	category, err := bamlreflect.Enum(ctx, "Category", []bamlreflect.EnumValue{
		bamlreflect.NewEnumValue("RED", bamlreflect.WithAlias("k7")),
	})
	if err != nil {
		t.Fatal(err)
	}
	record, err := bamlreflect.Class(ctx, "RuntimeRecord", []bamlreflect.Field{
		bamlreflect.NewField("label", bamlreflect.TypeOf[string](), bamlreflect.WithAlias("display_label")),
		bamlreflect.NewField("category", category),
		bamlreflect.NewField("scores", bamlreflect.TypeOf[int64]().Array()),
	})
	if err != nil {
		t.Fatal(err)
	}
	got, err := baml_sdk.HostParse[any](
		ctx,
		`{"display_label":"ok","category":"k7","scores":[1,2]}`,
		baml_go.WithTypeArg("T", record),
	)
	want := map[string]any{
		"label": "ok", "category": "RED", "scores": []any{int64(1), int64(2)},
	}
	if err != nil || !stdreflect.DeepEqual(got, want) {
		t.Fatalf("runtime class parse = %#v, %v; want %#v", got, err, want)
	}
}

func Test_compiled_package_returns_class_graph(t *testing.T) {
	ctx := context.Background()
	pkg, err := bamlreflect.CompilePackage(ctx, map[string]string{
		"runtime.baml": "class CompiledRow { amount int note string? }",
	})
	if err != nil {
		t.Fatal(err)
	}
	compiled, err := pkg.GetClass("CompiledRow")
	if err != nil || compiled == nil {
		t.Fatalf("get compiled class = %#v, %v", compiled, err)
	}
	got, err := baml_sdk.HostParse[any](ctx, `{"amount":7}`, baml_go.WithTypeArg("T", *compiled))
	want := map[string]any{"amount": int64(7), "note": nil}
	if err != nil || !stdreflect.DeepEqual(got, want) {
		t.Fatalf("compiled class parse = %#v, %v; want %#v", got, err, want)
	}
}

func Test_wire_occurrences_are_fresh_and_handles_reject_serialization(t *testing.T) {
	ctx := context.Background()
	runtimeType, err := bamlreflect.Class(ctx, "Fresh", []bamlreflect.Field{
		bamlreflect.NewField("value", bamlreflect.TypeOf[int64]()),
	})
	if err != nil {
		t.Fatal(err)
	}
	equal, err := baml_sdk.HostTypeEqual[any, any](
		ctx,
		baml_go.WithTypeArg("A", runtimeType),
		baml_go.WithTypeArg("B", runtimeType),
	)
	if err != nil || equal {
		t.Fatalf("fresh wire definitions equal = %v, %v; want false", equal, err)
	}
	if _, err := json.Marshal(runtimeType); err == nil || !strings.Contains(err.Error(), "cannot be serialized") {
		t.Fatalf("json serialization error = %v", err)
	}
	var encoded bytes.Buffer
	if err := gob.NewEncoder(&encoded).Encode(runtimeType); err == nil || !strings.Contains(err.Error(), "cannot be serialized") {
		t.Fatalf("gob serialization error = %v", err)
	}
}

func Test_known_type_tokens_compose_and_reject_unknowns(t *testing.T) {
	ctx := context.Background()
	assertTypeName := func(want string, ty bamlreflect.Type) {
		t.Helper()
		got, err := baml_sdk.HostTypeName[any](ctx, baml_go.WithTypeArg("T", ty))
		if err != nil || got != want {
			t.Fatalf("type name = %q, %v; want %q", got, err, want)
		}
	}

	assertTypeName("string", bamlreflect.TypeOf[string]())
	assertTypeName("(string | null)[]", bamlreflect.TypeOf[string]().Optional().Array())
	assertTypeName("StaticRecord", bamlreflect.TypeOf[baml_sdk.StaticRecord]())
	assertTypeName("StaticChoice", bamlreflect.TypeOf[baml_sdk.StaticChoice]())
	assertTypeName("StaticNamed", baml_sdk.StaticNamed)
	assertTypeName("image", bamlreflect.TypeOf[baml_go.Image]())

	type hostOnly struct{ Value string }
	_, err := baml_sdk.HostTypeName[any](ctx, baml_go.WithTypeArg("T", bamlreflect.TypeOf[hostOnly]()))
	if err == nil || !strings.Contains(err.Error(), "unsupported Go type token") {
		t.Fatalf("unsupported host token error = %v", err)
	}
}

func Test_host_handles_expose_composition_only(t *testing.T) {
	composed := bamlreflect.TypeOf[int64]().Meta(bamlreflect.WithDescription("count")).Type.Optional().Array()
	if _, err := json.Marshal(composed); err == nil {
		t.Fatal("composed type unexpectedly serialized")
	}
}

func Test_reflection_compile_errors_are_typed(t *testing.T) {
	_, err := bamlreflect.CompilePackage(context.Background(), map[string]string{
		"broken.baml": "class {",
	})
	var compilation *bamlreflect.CompilationError
	if !errors.As(err, &compilation) {
		t.Fatalf("compile error = %T %v; want *reflect.CompilationError", err, err)
	}
	if compilation.ClassName != "baml.reflect.errors.CompilationError" {
		t.Fatalf("compilation class = %q", compilation.ClassName)
	}
}
